lib/Jsrt/ChakraCore.h

```diff
@@ -104,6 +104,53 @@ typedef JsErrorCode(CHAKRA_CALLBACK * FetchImportedModuleFromScriptCallBack)(_In
 /// </returns>
 typedef JsErrorCode(CHAKRA_CALLBACK * NotifyModuleReadyCallback)(_In_opt_ JsModuleRecord referencingModule, _In_opt_ JsValueRef exceptionVar);
 
+/// <summary>
+///     A structure containing information about a native function callback.
+/// </summary>
+struct JsNativeFunctionInfo
+{
+    JsValueRef thisArg;
+    JsValueRef newTargetArg;
+    bool isConstructCall;
+};
+
+/// <summary>
+///     A function callback.
+/// </summary>
+/// <param name="callee">
+///     A function object that represents the function being invoked.
+/// </param>
+/// <param name="arguments">The arguments to the call.</param>
+/// <param name="argumentCount">The number of arguments.</param>
+/// <param name="info">Additional information about this function call.</param>
+/// <param name="callbackState">
+///     The state passed to <c>JsCreateFunction</c>.
+/// </param>
+/// <returns>The result of the call, if any.</returns>
+typedef _Ret_maybenull_ JsValueRef(CHAKRA_CALLBACK * JsEnhancedNativeFunction)(_In_ JsValueRef callee, _In_ JsValueRef *arguments, _In_ unsigned short argumentCount, _In_ JsNativeFunctionInfo *info, _In_opt_ void *callbackState);
+
+/// <summary>
+///     Creates a new enhanced JavaScript function.
+/// </summary>
+/// <remarks>
+///     Requires an active script context.
+/// </remarks>
+/// <param name="nativeFunction">The method to call when the function is invoked.</param>
+/// <param name="metadata">If this is a string, it is used as the name of the function.</param>
```

Comment:
What if this is not a string? If it must be a string (or undefined/not provided) perhaps name it `functionName` instead of `metadata`?

Reply:
For now it's acting as a stand-in for the name parameter from `JsCreateNamedFunction` but the idea is to extend it so that it could be a `Var` with properties like `length`, `name`, or `constructor`.

Reply:
Should we enforce the shape to be an Object with a single `name` property right now to make upgrading this easier in the future?